### PR TITLE
[SPARK-30994][BUILD][FOLLOW-UP] Change scope of xml-apis to include it and add xerces in SBT as dependency override

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR propose

1. Explicitly include xml-apis. xml-apis is already the part of xerces 2.12.0 (https://repo1.maven.org/maven2/xerces/xercesImpl/2.12.0/xercesImpl-2.12.0.pom). However, we're excluding it by setting `scope` to `test`. This seems causing `spark-shell`, built from Maven, to fail.

    Seems like previously xml-apis wasn't reached for some reasons but after we upgrade, it seems requiring. Therefore, this PR proposes to include it.

2. Pins `xerces` version in SBT as well. Seems this dependency is resolved differently from Maven.

Note that Hadoop 3 does not looks requiring this as they replaced xerces as of [HDFS-12221](https://issues.apache.org/jira/browse/HDFS-12221).

### Why are the changes needed?

To make `spark-shell` working from Maven build, and uses the same xerces version.

### Does this PR introduce any user-facing change?

No, it's master only.

### How was this patch tested?

**1.**

```bash
./build/mvn -DskipTests -Psparkr -Phive clean package
./bin/spark-shell
```

Before:

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/w3c/dom/ElementTraversal
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.apache.xerces.parsers.AbstractDOMParser.startDocument(Unknown Source)
	at org.apache.xerces.xinclude.XIncludeHandler.startDocument(Unknown Source)
	at org.apache.xerces.impl.dtd.XMLDTDValidator.startDocument(Unknown Source)
	at org.apache.xerces.impl.XMLDocumentScannerImpl.startEntity(Unknown Source)
	at org.apache.xerces.impl.XMLVersionDetector.startDocumentParsing(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
	at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
	at org.apache.xerces.parsers.DOMParser.parse(Unknown Source)
	at org.apache.xerces.jaxp.DocumentBuilderImpl.parse(Unknown Source)
	at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:150)
	at org.apache.hadoop.conf.Configuration.parse(Configuration.java:2482)
	at org.apache.hadoop.conf.Configuration.parse(Configuration.java:2470)
	at org.apache.hadoop.conf.Configuration.loadResource(Configuration.java:2541)
	at org.apache.hadoop.conf.Configuration.loadResources(Configuration.java:2494)
	at org.apache.hadoop.conf.Configuration.getProps(Configuration.java:2407)
	at org.apache.hadoop.conf.Configuration.set(Configuration.java:1143)
	at org.apache.hadoop.conf.Configuration.set(Configuration.java:1115)
	at org.apache.spark.deploy.SparkHadoopUtil$.org$apache$spark$deploy$SparkHadoopUtil$$appendS3AndSparkHadoopHiveConfigurations(SparkHadoopUtil.scala:456)
	at org.apache.spark.deploy.SparkHadoopUtil$.newConfiguration(SparkHadoopUtil.scala:427)
	at org.apache.spark.deploy.SparkSubmit.$anonfun$prepareSubmitEnvironment$2(SparkSubmit.scala:342)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.deploy.SparkSubmit.prepareSubmitEnvironment(SparkSubmit.scala:342)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:871)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1007)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1016)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.ClassNotFoundException: org.w3c.dom.ElementTraversal
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 42 more
```

After:

```
...
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.1.0-SNAPSHOT
      /_/

Using Scala version 2.12.10 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_202)
Type in expressions to have them evaluated.
Type :help for more information.

scala>
```

**2.**

```
./build/sbt dependencyTree -Phadoop-2.7 -Phive-2.3 -Phive-thriftserver -Phive
./build/sbt dependencyTree -Phadoop-3.2 -Phive-2.3 -Phive-thriftserver -Phive
```

### CODE REVIEW

**Comment by srowen** [comment]
Aha, OK that confirms https://issues.apache.org/jira/browse/SPARK-31043 (CC @tgravescs )
I think we also need https://github.com/apache/spark/pull/27790 for sbt, but will have to check the status of xml-apis there too to be sure.

I think this can likewise only be applied for Hadoop 2 profiles, but won't matter too much either way.

Another view here is that this isn't worth it for updating xerces, and just revert that. There is a possible security issue there, but not clear it matters in Spark.

WDYT?

**Comment by HyukjinKwon** [comment]
Oh, hm. I see. Sure, reverting will work too.
One thing we specify xerces at https://github.com/apache/spark/blob/master/project/SparkBuild.scala#L621-L625
Let me quickly try. If this doesn't work, we can just simply revert I guess.

**Comment by srowen** [comment]
You may be correct that xml-apis still needs to be included explicitly. However it does seem like it's already a compile dependency of xerces. Maybe your fix (or mine but yours is simpler) suffices to get it into the build for SBT transitively?

@tgravescs how did you build BTW? if it's SBT I get it; if it's Maven I'm a little puzzled.

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -621,6 +621,7 @@ object KubernetesIntegrationTests {
 object DependencyOverrides {
   lazy val settings = Seq(
     dependencyOverrides += "com.google.guava" % "guava" % "14.0.1",
+    dependencyOverrides += "xerces" % "xercesImpl" % "2.12.0",
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala]: SBT seems picking xml-apis and xerces fine when hadoop-2.7 profile is used:

```
[info]   | | | | +-xerces:xercesImpl:2.12.0
[info]   | | | | | +-xml-apis:xml-apis:1.4.01
```

Comment (HyukjinKwon) [reply project/SparkBuild.scala]: and .. not picked when hadoop-3.2 profile is used. If we concern about including `xml-apis`, we can just revert `xerces` too. I will leave it to you @srowen - I guess you have better insights on this.

File: pom.xml [hunk 2]
```diff
@@ -612,7 +612,6 @@
         <groupId>xml-apis</groupId>
         <artifactId>xml-apis</artifactId>
         <version>1.4.01</version>
-        <scope>test</scope>
```
Comment (HyukjinKwon) [inline pom.xml]: @srowen, I was wondering this actually overrides the scope of `xml-apis`, forces to be present in `test` scope only, and make it not present in the build output.

**Comment by HyukjinKwon** [comment]
@srowen, I added back the Maven change. Seems it is required.

**Comment by tgravescs** [comment]
I was building with maven and it is required, I dumped out dependency tree and xml-apis is a test dependency

**Comment by srowen** [comment]
Seems reasonable if it passes tests and gets all the expected dependencies.

We could also only make this happen for Hadoop 2. It's not needed in Hadoop 3. Then again maybe some value in providing xerces manually to harmonize the Spark env across both for apps, though I am always uneasy about providing more in the classpath.

There's also some argument that this isn't a simple update and the benefit is hard to quantify. I'd be OK if someone just wanted to abandon this update.

**Comment by srowen** [comment]
Hm, that's weird:

test.org.apache.spark.sql.JavaBeanDeserializationSuite.testSpark22000
```
java.lang.AssertionError: expected:<[JavaBeanDeserializationSuite.RecordSpark22000[shortField=0,intField=0,longField=0,floatField=0.0,doubleField=0.0,stringField=0,booleanField=true,timestampField=2020-03-05 07:31:44.0,nullIntField=<null>], 
...
...
but was:<[JavaBeanDeserializationSuite.RecordSpark22000[shortField=0,intField=0,longField=0,floatField=0.0,doubleField=0.0,stringField=0,booleanField=true,timestampField=2020-03-05 07:31:44,nullIntField=<null>], 
...
```

Difference is in the timestamp field. I don't see how it would be related to XML or Xerces; looking at whether recent master changes/reverts might have done this.

**Review by tgravescs — APPROVED** [review]
changes seem ok. I'm a bit surprised xml-apis not picked up from dependency in maven like it seems to be in sbt.  
Would be nice to get a fix committed and we can always follow up.

File: project/SparkBuild.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply project/SparkBuild.scala]: BTW, for `hadoop-3.2`, `xercesImpl` itself is not used, isn't it?

**Comment by holdenk** [comment]
Just a +1 on running into this as well

File: project/SparkBuild.scala (hunk 1 above)
Comment (HyukjinKwon) [reply project/SparkBuild.scala]: +1, Ideally yes, I think so.

**Comment by HyukjinKwon** [comment]
Thanks guys, let me merge this in.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.0.

**Comment by srowen** [comment]
Thanks everyone! on the one hand I think not catching it was my original PR's error, but getting that right wouldn't have surfaced this in tests, so thanks to @tgravescs for catching the issue downstream. Thanks @HyukjinKwon for the fix. I don't think we'll attempt this in 2.x unless there's demand.

**Pull request closed without merging**